File: julia/1_sonar_sweep/solution_DVH.jl
#=
author: Daan Van Hauwermeiren
AoC: day 1
=#
using Pkg
Pkg.activate(".")
using DelimitedFiles

fn = "input_day1"
# read data to vector
data = vec(readdlm(fn, Int64))

# function to get the number of positive values in a vector 
get_n_positive(a::Vector) = a .|> (x -> x>0) |> sum

solution_1 = get_n_positive(diff(data)) 

window_3_sum(x::Vector) = [sum(x[i-2:i]) for i in 3:length(x)]
solution_2 = get_n_positive(diff(window_3_sum(data))) 

File: python/1_sonar_sweep/solution_DVH.py
"""
author: Daan Van Hauwermeiren
AoC: day 1
"""

# %%
fn = "input_day1"

with open(fn, mode="r") as f:
    raw = f.readlines()
# %%
import numpy as np
def parse_input(data):
    return np.array([int(i.strip()) for i in data])
# %%
data = parse_input(raw)
# %%
solution_1 = sum(np.diff(data) > 0)
# %%
rolling_mean_3 = np.sum(
    np.vstack((
        np.roll(data, 0), 
        np.roll(data, 1), 
        np.roll(data, 2))),
    axis=0)[2:]
# %%
solution_2 = sum(np.diff(rolling_mean_3) > 0)
